File: packages/sdk-go/mcp_server.go
package rateguard

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
)

// ── MCP stdio server — zero-dependency JSON-RPC 2.0 transport ──
//
// ServeMCP speaks the Model Context Protocol stdio transport: newline-delimited
// JSON-RPC 2.0 messages on stdin/stdout. Any MCP client (Claude Code, Claude
// Desktop, Cursor, custom agents) can connect RateGuard as a tool server:
//
//	{"mcpServers": {"rateguard": {"command": "your-app", "args": ["mcp"]}}}
//
// Spec: https://modelcontextprotocol.io/specification/2025-06-18
// Methods implemented: initialize, notifications/initialized, ping,
// tools/list, tools/call. Everything else returns -32601 (method not found).

const (
	mcpProtocolVersion = "2025-06-18"
	mcpServerName      = "rateguard"

	jsonrpcParseError     = -32700
	jsonrpcInvalidRequest = -32600
	jsonrpcMethodNotFound = -32601
	jsonrpcInvalidParams  = -32602
)

type jsonrpcRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

type jsonrpcResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Result  any             `json:"result,omitempty"`
	Error   *jsonrpcError   `json:"error,omitempty"`
}

type jsonrpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

type mcpToolDescriptor struct {
	Name        string         `json:"name"`
	Description string         `json:"description"`
	InputSchema map[string]any `json:"inputSchema"`
}

type mcpCallParams struct {
	Name      string         `json:"name"`
	Arguments map[string]any `json:"arguments"`
}

// ServeMCP runs an MCP stdio server over the given reader/writer until the
// reader closes or ctx is canceled. Pass os.Stdin/os.Stdout to expose the
// SDK's pre-flight tools to any MCP client.
func (s *SDK) ServeMCP(ctx context.Context, r io.Reader, w io.Writer) error {
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	encoder := json.NewEncoder(w)

	for scanner.Scan() {
		if ctx != nil {
			select {
			case <-ctx.Done():
				return ctx.Err()
			default:
			}
		}

		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		var req jsonrpcRequest
		if err := json.Unmarshal(line, &req); err != nil {
			writeJSONRPC(encoder, jsonrpcResponse{
				JSONRPC: "2.0",
				Error:   &jsonrpcError{Code: jsonrpcParseError, Message: "parse error"},
			})
			continue
		}

		response, respond := s.handleMCPRequest(req)
		if respond {
			writeJSONRPC(encoder, response)
		}
	}

	return scanner.Err()
}

// handleMCPRequest dispatches one JSON-RPC message. respond is false for
// notifications (no id), which must not produce a response.
func (s *SDK) handleMCPRequest(req jsonrpcRequest) (jsonrpcResponse, bool) {
	isNotification := len(req.ID) == 0

	switch req.Method {
	case "initialize":
		return jsonrpcResponse{
			JSONRPC: "2.0",
			ID:      req.ID,
			Result: map[string]any{
				"protocolVersion": mcpProtocolVersion,
				"capabilities":    map[string]any{"tools": map[string]any{}},
				"serverInfo": map[string]any{
					"name":    mcpServerName,
					"version": Version,
				},
			},
		}, !isNotification

	case "notifications/initialized", "notifications/cancelled":
		return jsonrpcResponse{}, false

	case "ping":
		return jsonrpcResponse{JSONRPC: "2.0", ID: req.ID, Result: map[string]any{}}, !isNotification

	case "tools/list":
		descriptors := make([]mcpToolDescriptor, 0, 8)
		for _, tool := range s.MCPTools() {
			descriptors = append(descriptors, mcpToolDescriptor{
				Name:        tool.Name,
				Description: tool.Description,
				InputSchema: tool.InputSchema,
			})
		}
		return jsonrpcResponse{
			JSONRPC: "2.0",
			ID:      req.ID,
			Result:  map[string]any{"tools": descriptors},
		}, !isNotification

	case "tools/call":
		var params mcpCallParams
		if err := json.Unmarshal(req.Params, &params); err != nil || params.Name == "" {
			return jsonrpcResponse{
				JSONRPC: "2.0",
				ID:      req.ID,
				Error:   &jsonrpcError{Code: jsonrpcInvalidParams, Message: "tools/call requires params.name"},
			}, !isNotification
		}

		result, err := s.MCPCall(params.Name, params.Arguments)
		if err != nil {
			// Tool-level failures are reported in-band per the MCP spec.
			return jsonrpcResponse{
				JSONRPC: "2.0",
				ID:      req.ID,
				Result: map[string]any{
					"content": []map[string]any{{"type": "text", "text": err.Error()}},
					"isError": true,
				},
			}, !isNotification
		}
		return jsonrpcResponse{
			JSONRPC: "2.0",
			ID:      req.ID,
			Result:  map[string]any{"content": result.Content, "isError": false},
		}, !isNotification

	default:
		if isNotification {
			return jsonrpcResponse{}, false
		}
		return jsonrpcResponse{
			JSONRPC: "2.0",
			ID:      req.ID,
			Error:   &jsonrpcError{Code: jsonrpcMethodNotFound, Message: fmt.Sprintf("method not found: %s", req.Method)},
		}, true
	}
}

func writeJSONRPC(encoder *json.Encoder, response jsonrpcResponse) {
	if err := encoder.Encode(response); err != nil {
		log.Printf("rateguard: write mcp response: %v", err)
	}
}

File: packages/sdk-node/src/core/mcp-server.ts
/**
 * MCP stdio server — zero-dependency newline-delimited JSON-RPC 2.0.
 * Node port of Go's mcp_server.go (packages/sdk-go).
 *
 * serveMCP speaks the Model Context Protocol stdio transport: newline-
 * delimited JSON-RPC 2.0 messages on stdin/stdout. Any MCP client (Claude
 * Code, Claude Desktop, Cursor, custom agents) can connect RateGuard as a
 * tool server:
 *
 *   {"mcpServers": {"rateguard": {"command": "node", "args": ["your-mcp-entry.js"]}}}
 *
 * Spec: https://modelcontextprotocol.io/specification/2025-06-18
 * Methods implemented: initialize, notifications/initialized,
 * notifications/cancelled, ping, tools/list, tools/call. Everything else
 * returns -32601 (method not found). Notifications (id-less messages) never
 * get responses — including unknown notification methods.
 */

import { createInterface } from 'node:readline';

import type { MCPTool, MCPToolResult } from './mcp.js';

export const MCP_PROTOCOL_VERSION = '2025-06-18';
const MCP_SERVER_NAME = 'rateguard';

export const JSONRPC_PARSE_ERROR = -32700;
export const JSONRPC_METHOD_NOT_FOUND = -32601;
export const JSONRPC_INVALID_PARAMS = -32602;

interface JSONRPCRequest {
  jsonrpc?: string;
  id?: unknown;
  method?: string;
  params?: unknown;
}

interface JSONRPCResponse {
  jsonrpc: '2.0';
  id?: unknown;
  result?: unknown;
  error?: { code: number; message: string };
}

/**
 * The slice of the RateGuard facade the server needs — structural so the
 * server can be driven by the RateGuard class or anything shaped like it,
 * and shares that instance's limiter/budget/breaker state.
 */
export interface MCPServerHost {
  mcpTools(): MCPTool[];
  mcpCall(toolName: string, args?: Record<string, unknown>): Promise<MCPToolResult>;
}

/**
 * Runs an MCP stdio server over the given streams until the input closes.
 * Defaults to process.stdin/process.stdout; both are injectable so tests
 * (and embedders) can pipe arbitrary streams through it. Responses are
 * written strictly in request order even though tool handlers are async.
 */
export function serveMCP(
  host: MCPServerHost,
  stdin: NodeJS.ReadableStream = process.stdin,
  stdout: NodeJS.WritableStream = process.stdout,
): Promise<void> {
  const lines = createInterface({ input: stdin, crlfDelay: Infinity });

  const write = (response: JSONRPCResponse): void => {
    stdout.write(JSON.stringify(response) + '\n');
  };

  const handleLine = async (line: string): Promise<void> => {
    const trimmed = line.trim();
    if (trimmed.length === 0) {
      return;
    }

    let request: JSONRPCRequest;
    try {
      request = JSON.parse(trimmed) as JSONRPCRequest;
    } catch {
      write({ jsonrpc: '2.0', error: { code: JSONRPC_PARSE_ERROR, message: 'parse error' } });
      return;
    }

    const response = await handleMCPRequest(host, request);
    if (response) {
      write(response);
    }
  };

  return new Promise((resolve, reject) => {
    // Serialize handling so responses keep arriving in request order —
    // readline keeps emitting 'line' while an async handler is in flight.
    let queue: Promise<void> = Promise.resolve();
    lines.on('line', (line) => {
      queue = queue.then(() => handleLine(line));
    });
    lines.on('close', () => {
      queue.then(resolve, reject);
    });
  });
}

/**
 * Dispatches one JSON-RPC message. Returns undefined for notifications (no
 * id), which must not produce a response — including unknown notification
 * methods.
 */
async function handleMCPRequest(host: MCPServerHost, request: JSONRPCRequest): Promise<JSONRPCResponse | undefined> {
  const isNotification = request.id === undefined;

  switch (request.method) {
    case 'initialize':
      if (isNotification) return undefined;
      return {
        jsonrpc: '2.0',
        id: request.id,
        result: {
          protocolVersion: MCP_PROTOCOL_VERSION,
          capabilities: { tools: {} },
          serverInfo: {
            name: MCP_SERVER_NAME,
            version: process.env.RATEGUARD_VERSION || 'dev',
          },
        },
      };

    case 'notifications/initialized':
    case 'notifications/cancelled':
      return undefined;

    case 'ping':
      if (isNotification) return undefined;
      return { jsonrpc: '2.0', id: request.id, result: {} };

    case 'tools/list': {
      if (isNotification) return undefined;
      const tools = host.mcpTools().map((tool) => ({
        name: tool.name,
        description: tool.description,
        inputSchema: tool.inputSchema,
      }));
      return { jsonrpc: '2.0', id: request.id, result: { tools } };
    }

    case 'tools/call': {
      if (isNotification) return undefined;
      const params = (request.params ?? {}) as { name?: unknown; arguments?: unknown };
      const name = typeof params.name === 'string' ? params.name : '';
      if (!name) {
        return {
          jsonrpc: '2.0',
          id: request.id,
          error: { code: JSONRPC_INVALID_PARAMS, message: 'tools/call requires params.name' },
        };
      }

      const args =
        params.arguments && typeof params.arguments === 'object' && !Array.isArray(params.arguments)
          ? (params.arguments as Record<string, unknown>)
          : {};
      try {
        const result = await host.mcpCall(name, args);
        return {
          jsonrpc: '2.0',
          id: request.id,
          result: { content: result.content, isError: false },
        };
      } catch (error) {
        // Tool-level failures are reported in-band per the MCP spec, not as
        // JSON-RPC protocol errors.
        return {
          jsonrpc: '2.0',
          id: request.id,
          result: {
            content: [{ type: 'text', text: (error as Error).message }],
            isError: true,
          },
        };
      }
    }

    default:
      if (isNotification) {
        return undefined;
      }
      return {
        jsonrpc: '2.0',
        id: request.id,
        error: { code: JSONRPC_METHOD_NOT_FOUND, message: `method not found: ${request.method ?? ''}` },
      };
  }
}
